### PR TITLE
Bug 1481281 - Benchmark `Dataset.toJValue` against `Dataset.fieldsAsMap`

### PR DESCRIPTION
**Pull request by acmiyaguchi**
[buglink](https://bugzilla.mozilla.org/show_bug.cgi?id=1481281)

### CODE REVIEW

**Comment by acmiyaguchi** [comment]
@akkomar Do you have any comments or suggestions for this particular benchmark? I don't think there are any broad conclusions that can be made other than profiling on a job to job basis to check that garbage collection isn't a large portion of the jobs time.

File: moztelemetry/bug-1481281-message-jvalue/README.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+#### fieldsAsMap with G1
+![map-cms](images/map-g1.png)
+
+G1 with the Map method has a smaller sawtooth period than the the JValue method. The garbage collector more
+frequently clears the heap of unused objects.
+
+## Discussion
+
+There are no clear performance wins for the `Message.toJValue` method in the microbenchmarks. It's unclear what
```
Comment (akkomar) [inline moztelemetry/bug-1481281-message-jvalue/README.md]: How do microbenchmarks' the input data size compare to daily MainSummary runs?

**Comment by akkomar** [comment]
@acmiyaguchi I agree. Couple of comments:
* I noticed that previously (with `fieldsAsMap`) there was a lot of typecasting in `MainSummaryView` which was removed thanks to inspecting field types. Is that correct? I think this might be one of the reasons for the improvement. On a second it looks like you're doing this typecasting in benchmark too (`flattenTreeMap`), so the mystery remains...
* Another difference between compared versions of `MainSummaryView` is extensive usage of lazy vals in the old one (https://github.com/mozilla/telemetry-batch-view/blob/bb1abf1cf631f5263ade0c2f4aab0b4ac4831dc3/src/main/scala/com/mozilla/telemetry/views/MainSummaryView.scala#L419-L429). Its removal might contribute to the speedup as they're synchronized under the hood
* Interesting follow up to this might be doing method duration profiling (this profiler looks promising for Spark applications: https://github.com/uber-common/jvm-profiler).

**Comment by erindcole** [comment]
Your findings are well described. It would be even more informative to add more specifics in the `Discussion` section. I'd like to see each benchmark file mentioned in order of how it is introduced in the `Results` section. One sentence that describes how each benchmark adds impact in the `Discussion` specifically would conclude the findings well.

**Review by erindorothy — APPROVED** [review]
Your findings are well described. It would be even more informative to add more specifics in the `Discussion` section. I'd like to see each benchmark file mentioned in order of how it is introduced in the `Results` section. One sentence that describes how each benchmark adds impact in the `Discussion` specifically would conclude the findings well.

**Comment by acmiyaguchi** [comment]
@akkomar 

* There's still quite a bit of type-casting, but it's been moved over to JValue's built-in `.extractOpt[T]`. 
* Lazy evaluation definitely could have been a factor here. Looking at the [code that's generated by the compiler](http://scala-programming-language.1934581.n4.nabble.com/scala-Lazy-val-and-synchronized-td1991409.html), the book keeping overhead could have been contributing to run-time. I'm not sure that this would be triggering the garbage collector more often though.
* I'll make sure to add a follow-up section, I think profiling a job like longitudinal might be really interesting.

@erindcole I'll make sure to weave in the benchmarks into the discussion. It's unfortunate that the results don't lead to any strong conclusions, but I can expand more on future direction.

**Comment by acmiyaguchi** [comment]
I've added to the discussion by expanding on the results of the benchmarks and some of the related code changes that were made. I also point to some future profiling that could be done on our scala code near the end. Thanks @akkomar and @erindcole for taking the time to read and comment.

**Pull request merged**